### DISCUSSION TITLE
Grab focus when new buffer is opened

### DISCUSSION THREAD

**Discussion by lsoksane** [orbitalquark/textadept#130] [Ideas]
I am using Textadept on Ubuntu. It would be great to have the application to grab focus when I open a file with it using a file manager or command line. Ideally I would like this to happen regardless whether the file I ask Textadept to open is already open in it or not. Just an event to connect to would be enough.

**Comment by orbitalquark** [comment]
On my Linux box opening files in Textadept via the terminal (LXTerminal) and file manager (PCManFM) always grabs focus. What terminal and file manager are you using? Does Textadept open the files, but simply remain in the background?

**Reply by lsoksane** [reply]
Textadept opens the files as expected but it remains in the background. I am using Caja (MATE caja 1.20.2), Mate Terminal (MATE Terminal 1.20.0) and Textadept 11.2. I did some further testing and the behaviour is quite strange. I tested this without any customizations in init.lua. 

When I first open a file, with Textadept not running, it gets focus. If I don't click the Textadept window, I can keep opening files from Caja and Textadept gets the focus as expected, regardless whether I open a new file or a file that is already open in Textadept. But if I click Textadept, then go back to Caja and open some more files, Textadept does not get focus anymore.

**Comment by orbitalquark** [comment]
It sounds as if perhaps your window manager is doing something odd. Unfortunately, I don't know what can be done here other than look to see if there is a window manager setting to turn off this feature.

**Comment by manuelmontenegro** [comment]
I have been able to reproduce this issue in Fedora 34 with GNOME 40. If I use Wayland, the TextAdept window does not grab focus when opening a file from Nautilus (aka GNOME Files). If I use Xorg instead, TextAdept does grab focus. Perhaps this is the difference...

It might be related to an issue with `gtk_window_present()` in Wayland: https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/258

I tried replacing the call to `gtk_window_present()` in [textadept.c:255](https://github.com/orbitalquark/textadept/blob/80816ecdba807af53f556dc3523b7726625bd5c4/src/textadept.c#L255) with `gtk_window_present_with_time()` passing `gtk_get_current_event_time()` as second parameter, and seemed to work under Wayland, but I am not an experienced Gtk developer, so please take this with a grain of salt :-)

**Reply by lsoksane** [reply]
Many thanks for looking into this. This is not an issue with Wayland per se, as I am using Xorg. I also tested with a fresh iso of Ubtuntu Mate 20.04.3 in VirtualBox. I can reproduce the issue in this straight out of the box setting.

**Comment by orbitalquark** [comment]
It may be worth building Textadept with `make GTK3=1` and seeing if it's a GTK 2 issue (Textadept binaries are compiled against GTK 2).

**Reply by lsoksane** [reply]
I did this on a freshly installed Ubtuntu Mate 20.04.3, and I still get the same behaviour.

**Comment by lsoksane** [comment]
A work-around solving the issue is to replace <https://github.com/orbitalquark/textadept/blob/80816ecdba807af53f556dc3523b7726625bd5c4/src/textadept.c#L255>
by 
```
  return (gtk_window_present_with_time(GTK_WINDOW(window), (g_get_monotonic_time() / 1000) & 0xFFFFFFFF), 0);
```
The supposedly more sane
```
  return (gtk_window_present_with_time(GTK_WINDOW(window), gtk_get_current_event_time()), 0);
```
does not work for me. Note that, regarding the present code using gtk_window_present, the documentation 
<https://docs.gtk.org/gtk3/method.Window.present.html>
says that 'This function should not be used'. My option is discussed further in <https://github.com/projecthamster/hamster/issues/555> where it is deemed dangerous. I just couldn't come up with anything else that works.

**Comment by orbitalquark** [comment]
Thanks for taking the time to look into this some more.

My local copy of the GTK documentation says "If you are calling [gtk_window_present()] in response to a user interaction, it is preferable to use gtk_window_present_with_time()." Since Textadept is not responding to a GTK event (with a GTK-generated timestamp), I am reluctant to manufacture an event time, and it seems like the people in the link you mentioned are too. If you find something else, let me know. I likely will not spend more time on this since it works for me and I don't have the means to test in another environment.
